Source/Dafny/Resolver.cs

```diff
@@ -13216,7 +13217,10 @@ private class RBranchExpr : RBranch {
         call = (asx.Lhs.Resolved as MemberSelectExpr)?.Member as Method;
         if (call != null) {
           // We're looking at a method call
-          if (call.Outs.Count != 0) {
+          var outTypes = (asx.Lhs.Resolved as MemberSelectExpr)?.ResolvedOutparameterTypes;
+          if (outTypes != null && outTypes.Count != 0) {
+            firstType = outTypes[0];
+          } else if (call.Outs.Count != 0) {
```

Comment:
The information you're storing in the new `ResolvedOutparameterTypes` is already available in the `MemberSelectExpr` node. Instead of the new lines 13220-13223, just set `firstType` as follows:

    var typeMap = ((MemberSelectExpr)asx.Lhs.Resolved).TypeArgumentSubstitutionsWithParents();
    firstType = SubstType(call.Outs[0].Type, typeMap);

You'll also need to fix another problem in `Translator.cs`, namely to move

    vdecl.Locals.Iter(RemoveDefiniteAssignmentTracker);

out from the `else` branch (so that it's done whether or not `vdecl.Update is AssignOrReturnStmt` holds) in method `RemoveDefiniteAssignmentTrackers`.

Reply:
The doc of `TypeArgumentSubstitutionsWithParents` says `NOTE: This method should be called only when all types have been fully and successfully resolved.`
We are not in the phase when all types have been successfully resolved, because we are still resolving symbols inside the AssignAndReturnStmt. Should I update the comment?

I did implement your suggested change.